File: core/src/main/scala/eu/easyminer/rdf/data/ops/Cacheable.scala
package eu.easyminer.rdf.data.ops

import java.io.{InputStream, OutputStream}

import eu.easyminer.rdf.utils.serialization.{Deserializer, SerializationSize, Serializer}

/**
  * Created by Vaclav Zeman on 27. 2. 2018.
  */
trait Cacheable[T, Coll] extends Transformable[T, Coll] {

  self: Coll =>

  protected implicit val serializer: Serializer[T]
  protected implicit val deserializer: Deserializer[T]
  protected implicit val serializationSize: SerializationSize[T]

  def cache: Coll = transform(coll.toList)

  def cache(os: => OutputStream, is: => InputStream): Coll = {
    Serializer.serializeToOutputStream[T](os) { writer =>
      coll.foreach(writer.write)
    }
    transform(new Traversable[T] {
      def foreach[U](f: T => U): Unit = {
        Deserializer.deserializeFromInputStream[T, Unit](is) { reader =>
          Stream.continually(reader.read()).takeWhile(_.isDefined).foreach(x => f(x.get))
        }
      }
    })
  }

  def cache(os: => OutputStream): Coll = {
    Serializer.serializeToOutputStream[T](os) { writer =>
      coll.foreach(writer.write)
    }
    this
  }

}

File: java/src/main/java/com/github/propi/rdfrules/java/Cacheable.java
package com.github.propi.rdfrules.java;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.function.Supplier;

/**
 * Created by Vaclav Zeman on 3. 5. 2018.
 */
public interface Cacheable<T, SColl, JColl> extends Transformable<T, SColl, JColl> {

    com.github.propi.rdfrules.data.ops.Cacheable<T, SColl> asScala();

    default JColl cache() {
        return asJava(asScala().cache());
    }

    default JColl cache(Supplier<OutputStream> osb) {
        return asJava(asScala().cache(osb::get));
    }

    default JColl cache(Supplier<OutputStream> osb, Supplier<InputStream> isb) {
        return asJava(asScala().cache(osb::get, isb::get));
    }

}
